docs/html/user_guide.rst

```diff
@@ -861,6 +861,296 @@ of ability. Some examples that you could consider include:
+      oauthlib==3.1.1 --hash sha256:42bf6354c2ed8c6acb54d971fce6f88193d97297e18602a3a886603f9d7730cc
+      pyasn1==0.4.8 --hash sha256:39c7e2ec30515947ff4e87fb6f456dfc6e84857d34be479c9d4a4ba4bf46aa5d
+      setuptools==60.3.1 --hash sha256:2932bfeb248c648dc411ea9714d5a6de7a33ef1a0db2f0fce644d8172b0479e8
+
+JSON schema
+-----------
+
+The JSON report is described below with fields given fake mypy type annotations corresponding to the output of `json.load() <https://docs.python.org/3/library/json.html#json.load>`_ upon reading the JSON report. The report contains multiple top-level fields:
+
+* ``experimental: bool``: set to ``True`` as the format is not yet stable.
+* ``input_requirements: List[str]``: strings describing the requirements provided to the pip resolver. Can be parsed with `packaging.requirements.Requirement <https://packaging.pypa.io/en/latest/requirements.html>`_.
+* ``python_version: str``: a string describing the python interpreter version the resolve was performed for. Can be parsed with `packaging.specifiers.SpecifierSet <https://packaging.pypa.io/en/latest/specifiers.html>`_. Currently always an exact ``==`` constraint.
+* ``candidates: Dict``: each package that would have been downloaded with ``pip download`` is represented in the ``candidates`` dict. Each key is the ``name`` of a requirement, since each dependency package name is satisfied by exactly one candidate in the final resolve.
+
+Candidates
+----------
+
+Each element of the ``candidates`` dict has the following fields:
+
+* ``requirement: str``: an ``==`` requirement string for the exact version of the candidate that would have been fetched by ``pip download``. Can be parsed with `packaging.requirements.Requirement <https://packaging.pypa.io/en/latest/requirements.html>`_.
+* ``requires_python: Optional[str]``: a constraint on the executing python interpreter version exerted by this candidate. Can be parsed with `packaging.specifiers.SpecifierSet <https://packaging.pypa.io/en/latest/specifiers.html>`_.
+* ``dependencies: List[str]``: all the dependencies required by this candidate, as requirement strings which can be parsed with `packaging.requirements.Requirement <https://packaging.pypa.io/en/latest/requirements.html>`_. Each requirement will have been satisfied by another member of the overall ``candidates`` dict.
+* ``download_info: Dict``: a link where the requirement can be directly downloaded from, along with any metadata.
+
+Download Info
+-------------
+
+The ``download_info`` object has the following fields:
+
+* ``direct_url: Dict``: a deserializable representation of a Direct URL as per :pep:`610` for this package's location, which may be a remote URL or local directory.
+* ``dist_info_metadata: Optional[Dict]``: a deserializable representation of a Direct URL as per :pep:`610` for this package's *metadata*, which may be provided for individual package downloads by a package index provided with ``-i`` which implements :pep:`658`.
```

Comment:
What is the use case for having this PEP 658 url in the report ?

I'd suggest instead adding a `metadata` field with the full metadata in JSON format (per PEP 566). This field would be present whatever the origin of the metadata (a wheel, PEP 658, or a local metadata preparation/build step).

Reply:
I we remove dist_info_metadata from download_info, then download_info can be the direct_url data structure.

Reply:
I totally agree with this, as per https://github.com/pypa/pip/pull/10748#issuecomment-1123307028. I think we should remove `dist_info_metadata` from there, allow `download_info` to just be the `direct_url` data structure, and instead providing the full json metadata in the `metadata` field.